src/core/processing/qgsprocessingalgorithm.h

```diff
@@ -697,6 +699,115 @@ Q_DECLARE_OPERATORS_FOR_FLAGS( QgsProcessingAlgorithm::Flags )
 
 
 
+/**
+ * \class QgsProcessingFeatureBasedAlgorithm
+ * \ingroup core
+ * An abstract QgsProcessingAlgorithm base class for processing algorithms which operate "feature-by-feature".
+ *
+ * Feature based algorithms are algorithms which operate on individual features in isolation. These
+ * are algorithms where one feature is output for each input feature, and the output feature result
+ * for each input feature is not dependent on any other features present in the source.
+ *
+ * For instance, algorithms like "centroids" and "buffers" are feature based algorithms since the centroid
+ * or buffer of a feature is calculated for each feature in isolation. An algorithm like "dissolve"
+ * is NOT suitable for a feature based algorithm as the dissolved output depends on multiple input features
+ * and these features cannot be processed in isolation.
+ *
+ * Using QgsProcessingFeatureBasedAlgorithm as the base class for feature based algorithms allows
+ * shortcutting much of the common algorithm code for handling iterating over sources and pushing
+ * features to output sinks. It also allows the algorithm execution to be optimised in future
+ * (for instance allowing automatic multi-thread processing of the algorithm, or use of the
+ * algorithm in "chains", avoiding the need for temporary outputs in multi-step models).
+ *
+ * \since QGIS 3.0
+ */
+
+class CORE_EXPORT QgsProcessingFeatureBasedAlgorithm : public QgsProcessingAlgorithm
+{
+  public:
+
+    /**
+      * Constructor for QgsProcessingFeatureBasedAlgorithm.
+      */
+    QgsProcessingFeatureBasedAlgorithm() = default;
+
+  protected:
+
+    void initAlgorithm( const QVariantMap &configuration = QVariantMap() ) override;
+
+    /**
+     * Returns the translated, user visible name for any layers created by this algorithm.
+     * This name will be used as the default name when loading the resultant layer into a
+     * QGIS project.
+     */
+    virtual QString outputName() const = 0;
+
+    /**
+     * Returns the layer type for layers generated by this algorithm, if
+     * this is possible to determine in advance.
+     */
+    virtual QgsProcessing::LayerType outputLayerType() const { return QgsProcessing::TypeVectorAny; }
+
+    /**
+     * Maps the input WKB geometry type (\a inputWkbType) to the corresponding
+     * output WKB type generated by the algorithm. The default behavior is that the algorithm maintains
+     * the same WKB type.
+     */
+    virtual QgsWkbTypes::Type outputWkbType( QgsWkbTypes::Type inputWkbType ) const { return inputWkbType; }
```

Comment:
What's that used for? Just wondering how that integrates with, say `createFromAttributeWkt` which also operates feature-by-feature.